### ISSUE TITLE
[Feature Request] Notification Cooldown

### ISSUE THREAD

**Issue by ChadBailey** [telegramdesktop/tdesktop#7618]
### Is your feature request related to a problem?
Everyone has a love hate relationship with notifications. We want to be notified, but we don't like being annoyed with constant notifications. How do we reach this balance?

### Describe the solution you'd like
This is a novel approach to notifications not found elsewhere which I will call notification cooldown. I invented this around 2002 and has been in use in my own unreleased software since. 

> IMPORTANT: I'm keenly aware that unproven notification changes are a very popular request, and this may at first glance seem to share similarities with the existing smart notifications feature. I ask that you please consider setting your expectations aside for a moment and hear me out. This has been proven to work well, and if you give it a try you may find as I did that it works even better than it sounds.

Here is a gherkin formatted feature to make the desired behavior crystal clear

```feature
Feature: Notification Cooldown

    As a user of the notification cooldown feature, I want to only be notified
    of new conversations so that I do not have to manually disable notifications
    to ignore conversations I am unable to participate in

    # Background is present in all scenarios
        Background:
            Given a telegram user "user1"
              And "user1" is part of group chats "group1" and "group2"
              And "user1" has notification cooldown enabled for "group1" and "group2"
              And the notification cooldown is set to 5 minutes

        Scenario: Notified of new message
            Given "group1" has not received a message within 5 minutes
             When a new message arrives
             Then "user1" is notified

        Scenario: Not notified of new message
            Given "group1" received a message within 5 minutes
             When a new message arrives
             Then "user1" is not notified

        Scenario: One notification over ten minutes
            Given "group1" received a message ("user1" notified)
              And 4 minutes later another message arrived ("user1" not notified)
              And 3 minutes later another message arrived ("user1" not notified)
              And 2 minutes later another message arrived ("user1" not notified)
             When "group1" receives a message 1 minute after the last message (10 minutes after initial message)
             Then "user1" is not notified

        Scenario: Two notifications in one minute
            Given "group1" received a message
              And "user1" received a notification
              And "user1" focused "group1"
              And "user1" then defocused "group1"
             When "group1" receives another message within 1 minute
             Then "user1" receives a notification for the new message

        Scenario: Notifications from multiple chats
            Given "group1" has not received a message within 5 minutes
              And "group2" has not received a message within 5 minutes
             When "group1" receives a new message
              And "group2" receives a new message 1 minute later
             Then "user1" is notified of the message in "group1"
              And "user1" is notified of the message in "group2"
```

### Describe alternatives you've considered
- Mute notifications entirely
- Mute notifications manually for a duration
- Smart notifications (1 time within 5 minutes)

### Additional context
This feature provides benefits that seem mutually exclusive, which is what makes it desirable
1. Back to back notifications are eliminated
2. Provides a guarantee that an unattended chat can notify at maximum once per 5 minutes, but typically far less often
3. Provides guarantee that attended chats receive notifications immediately
4. Does not require other chat participants change their behavior

You may be wondering why I would give this to Telegram freely if I regard it as something I have kept to myself for nearly 20 years. The truth is, Telegram is the chat client I enjoy using most and I don't see me making a chat client. My unreleased software is IRC based and was written only for me so I never bothered making it easy to use or configurable.

**Comment by ilya-fedin** [comment]
#429?

**Comment by ChadBailey** [comment]
i did my best to call this out very clearly in my report, but i will summarize here. while this seems similar to #429 , the outcomes are very different.

specifically, according to my understanding of the implementation there are 2 key differences
1. #429 will not eliminate notifications from continuous conversations lasting beyond threshold which.
  > example: for arguments sake say i have it set to "1 time within 5 minutes" and the conversation is indefinite.
  > 
  >
  > using #429 i will be notified every 5 minutes.
  >
  > using this method, i will be notified only 1 time. over the course of 1 hour, that is a difference of 20 more notifications

2. #429 will not guarantee i will be notified when i am participating in a conversation
  > example: i have it set to "1 time within 5 minutes" and i have received the message "hey, what's up?" i reply "nothing, how about you?" then continue with what i was doing before, defocusing telegram. the person then responds (within 1 minute of the first message) "i need you to call me urgently" then "hello, are you there?" then "where did you go?" "please, this is very urgent" "look i had a sweet gig lined up for you but he wanted to talk to you on the phone before he went through with the deal. maybe next time".
  >
  >
  > with #429 i would never be notified of any messages past "hey, what's up?".
  >
  > using this method i would be notified of "hey, what's up?", as well as "i need you to call me urgently"

**Comment by ChadBailey** [comment]
I have tried re-reading the request more carefully to be sure I wasn't missing anything...

I see that it is still open/never implemented. I can also see how this ambiguous feature request could be considered the same as mine, but the author did not really provide specifications to code to as I did so it's hard to say if they intended the same behavior.

They claimed the resolution to be _"Merge new line-wise part-messages if they are still in a certain time threshold (e.g. 5 seconds) with the old message. Do not trigger a new sound notification."_. I am not requesting merging any messages. With my requested behavior the source of the new messages could be a single user or multiple users all engaging in a conversation over a far longer period of time (5 minutes vs 5 seconds).

These differences seem small and insignificant, but they are not insignificant at all. Perhaps the reason this has been outstanding for 5 years is because no one has understood what the desired behavior actually was.
